lib/src/main/java/com/knuddels/jtokkit/TokenEncoder.java

```diff
@@ -1,80 +1,196 @@
 package com.knuddels.jtokkit;
 
+
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
-import java.util.Optional;
-import java.util.function.Function;
-
-/**
- * A TokenEncoder is used to encode and decode tokens. It is initialized with a map
- * that contains the decoded tokens as keys and the encoded tokens as values. The
- * TokenEncoder can then be used to encode and decode tokens.
- */
-class TokenEncoder {
-
-    private final Map<ImmutableByteArray, Integer> decodedToEncoded = new HashMap<>();
-
-    /**
-     * Creates a new TokenEncoder with the given input map. The keys of the map are
-     * the decoded tokens and the values are the encoded tokens.
-     *
-     * @param input the input map
-     */
-    public TokenEncoder(Map<ImmutableByteArray, Integer> input) {
-        this(input, Function.identity());
+import java.util.TreeMap;
+import java.util.concurrent.ConcurrentHashMap;
+
+final class TokenEncoder {
+    public static final int DUMMY_RANK = Integer.MAX_VALUE;
+    public static final int MAX_RANK = Integer.MAX_VALUE - 1;
+    private final Map<ImmutableByteArray, Integer>[] encoders;
+    private int length = 0;
+
+    public TokenEncoder(Map<byte[], Integer> encoder) {
+        if (!encoder.isEmpty()) {
+            TreeMap<Integer, Map<ImmutableByteArray, Integer>> tempEncoders = new TreeMap<>();
+            encoder.forEach((k, v) -> {
+                length++;
+                ImmutableByteArray key = ImmutableByteArray.from(k);
+                tempEncoders.computeIfAbsent(k.length, integer -> new ConcurrentHashMap<>()).put(key, v);
+            });
+            //noinspection unchecked
+            encoders = new ConcurrentHashMap[tempEncoders.lastKey() + 1];
+            tempEncoders.forEach((k, v) -> encoders[k] = v);
+        } else {
+            encoders = new HashMap[0]; // for testing
+        }
+    }
+
+    public static int getMinRankIndex(List<Integer> ranks) {
```

Comment:
When searching for the next minimum value, we've unrolled it to favor SIMD optimizations

Reply:
it is still left for the JIT compiler, since I'm not using the [vector api](https://openjdk.org/jeps/426), just making it simpler for it to group similar instructions. If you could run the benchmarks yourself, you can tell me yourself if it reproduces or not.
And yes, I have benchmarked everything separately (haven't committed every single benchmark though), this was the fastest.